### PR TITLE
Use DLSImages service to convert from mrc to jpeg for CTF diagnostic image and motion corrected micrograph

### PR DESCRIPTION
**Pull request by d-j-hatton**
To display Relion output images in SynchWeb we need to convert them from mrc format to jpeg and then insert the path into ISPyB. These changes should provide the infrastructure for doing that (on the wrapper side) for CTF diagnostic plots and motion corrected micrographs. 

The conversion will be done by the DLSImages zocalo service using the `images_service_plugin` provided here. This will require changes to the implementation of the Images service. Sending messages to the service will require a change to the zocalo recipe as well.

### CODE REVIEW

File: tests/parser/test_ctffind.py [hunk 1]
```diff
@@ -79,3 +80,14 @@ def test_invalid_job_references_raise_KeyErrors(ctffind):
 
 def test_astigmatism(ctffind):
     assert ctffind["job003"][0].astigmatism == "288.135742"
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="does not run on windows")
+def test_diagnosis_plot_path(proj, ctffind):
+    assert ctffind["job003"][0].diagnostic_plot_path == str(
+        pathlib.PurePosixPath(proj.basepath)
```
Comment (Anthchirp) [inline tests/parser/test_ctffind.py:88]: I believe `.basepath` is guaranteed to be a path object, so you shouldn't need to recast it?

File: src/relion/zocalo/wrapper.py [hunk 2]
```diff
@@ -328,6 +333,28 @@ def get_status(self, job_path):
...
+    logger.debug(f"{relion_stage_object!r} does not have associated images")
+    return []
+
+
+@images_msgs.register(relion.MotionCorr)
+def _(stage_object: relion.MotionCorr, job_string: str):
+    imgs = []
+    for micrograph in stage_object[job_string]:
+        imgs.append(micrograph.micrograph_snapshot_full_path.replace(".jpeg", ".mrc"))
+    return imgs
```
Comment (Anthchirp) [inline src/relion/zocalo/wrapper.py:347]: Recommend using list comprehensions where possible over appending elements in a loop
```suggestion
    return [
        micrograph.micrograph_snapshot_full_path.replace(".jpeg", ".mrc")
        for micrograph in stage_object[job_string]
    ]
```

File: tests/zocalo/test_images_service_plugin.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    ct.nack.assert_called_once()
+    ct.ack.assert_not_called()
+
+
+def test_mrc_to_jpeg_ack_when_file_exists(tmp_path):
+    jpeg_path = str(tmp_path / "convert_test.jpeg")
+
+    test_data = np.arange(9, dtype=np.int8).reshape(3, 3)
+    with mrcfile.new(jpeg_path.replace(".jpeg", ".mrc")) as mrc:
+        mrc.set_data(test_data)
```
Comment (Anthchirp) [inline tests/zocalo/test_images_service_plugin.py:66]: Would advise to keep the path object around, as it makes operations like changing extensions really easy:
```suggestion
    jpeg_path = tmp_path / "convert_test.jpeg"

    test_data = np.arange(9, dtype=np.int8).reshape(3, 3)
    with mrcfile.new(jpeg_path.with_suffix(".mrc")) as mrc:
        mrc.set_data(test_data)
```
and if `mrcfile.new()` can't deal with path objects, then wrap with `os.fspath()`. Uses of `jpeg_path` further down in the function would then also need updating.

File: requirements_dev.txt [hunk 4]
```diff
@@ -4,3 +4,8 @@ gemmi==0.4.6
 graphviz==0.16
 pytest==6.2.4
 pytest-cov==2.12.0
+mrcfile==1.3
+workflows==2.9
+pillow==8.0
+numpy==1.21; python_version >= '3.7'
+numpy==1.19; python_version < '3.7'
```
Comment (Anthchirp) [inline requirements_dev.txt:11]: These (except `workflows`) will need to be replicated in `requirements.txt`, possibly in `requirements_doc.txt` (we'll see on merging). Packages also need adding to `setup.cfg`'s `install_requires`, but without the version constraint.

File: tests/parser/test_ctffind.py (hunk 1 above)
Comment (d-j-hatton) [reply tests/parser/test_ctffind.py:88]: This is a hangover from when I was trying to get the test to work on Windows, although I ultimately gave up. I'll remove it

**Pull request merged**